### DISCUSSION TITLE
How to use sub folders in Supabase edge functions project?

### DISCUSSION THREAD

**Discussion by zachbresler21** [supabase/supabase#18441] [Questions]
I am attempting to organize various functions into specific folders for better structure and clarity. However, I'm encountering an issue where Supabase isn't recognizing the functions stored within these categorized directories. To illustrate, an example structure:

- fast-functions
    - function1
    - function2
- slow-functions
    - function1
    - function2

Despite this organization, the functions nested within either 'fast-functions' or 'slow-functions' aren't being detected as expected. I'm looking for a solution that maintains this logical division while ensuring seamless Supabase integration.

Currently all functions in the root of the /supabase/functions folder deploy and work perfectly but as my project grows I need to structure the functions. 

Would really appreciate some help here. Thanks

**Comment by shreyash0k** [comment]
same issue

**Reply by dacameron-proactive** [reply]
Did this ever get raised as an issue? It seems like table-stakes to me for anything other than a superficial project.

**Comment by Kilowhisky** [comment]
Code says that is a no-go right now.  (`*` vs `**`)

https://github.com/supabase/cli/blob/8f3034ef818d370ad3e75ff310fe72d476ab2133/internal/functions/deploy/deploy.go#L49-L66
 ```
func GetFunctionSlugs(fsys afero.Fs) (slugs []string, err error) {
	pattern := filepath.Join(utils.FunctionsDir, "*", "index.ts")
	paths, err := afero.Glob(fsys, pattern)
	if err != nil {
		return nil, errors.Errorf("failed to glob function slugs: %w", err)
	}
	for _, path := range paths {
		slug := filepath.Base(filepath.Dir(path))
		if utils.FuncSlugPattern.MatchString(slug) {
			slugs = append(slugs, slug)
		}
	}
	// Add all function slugs declared in config file
	for slug := range utils.Config.Functions {
		slugs = append(slugs, slug)
	}
	return slugs, nil
}
```

I would love sub-folder support too.

**Reply by craig-at-metashield-io** [reply]
yes, please! 

if you have any non-trivial number of functions (more than a dozen or so), it gets pretty hard to manage.

**Comment by beautyfree** [comment]
Has anything changed?

**Comment by craig-at-metashield-io** [comment]
what compounds this issue, btw, is that the edgefunctions page of the supabase webapp sorts the functions by create date, as opposed to alphabetical, which makes working with 30 or so functions pretty damn rough.

**Reply by rookledookle** [reply]
agree!

**Reply by dacameron-proactive** [reply]
@craig-at-metashield-io has this been raised as an issue? Seems like a) a prtty obvious UX bug; b) really easy for the SB team to fix.

If there's a ticket, I'd vote.

**Comment by pacocartones** [comment]
This is still intentionally limited to one directory level in the current CLI. Its automatic discovery pattern is [`supabase/functions/*/index.ts`](https://github.com/supabase/cli/blob/develop/apps/cli-go/internal/functions/deploy/deploy.go#L79-L80), so `supabase/functions/fast-functions/function1/index.ts` is not discovered as a deployable function.

The supported structure is therefore to keep each deployable function directly under `functions`, and organize *inside* it (or in `_shared`):

```text
supabase/functions/
  _shared/
    auth.ts
  fast-function1/
    index.ts
    lib/
  slow-function1/
    index.ts
    lib/
```

Use names such as `fast-function1` / `slow-function1` to preserve the grouping in the dashboard and CLI. The current docs likewise recommend co-locating a function's own modules with its `index.ts`, moving only cross-function code to `_shared`: https://supabase.com/docs/guides/functions/unit-test#recommended-project-structure

So there is no recursive folder discovery to enable at present; changing that would need a CLI feature rather than a project-layout setting.
